File: c++/convertZigZag.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <cmath>
using namespace std;

string convert(string s, int nRows) {
    if (s.length() == 0 || s.length() <= nRows || nRows == 1) return s;

    string result = "";
    int zig = 2 * (nRows - 1);
    int zag = zig;

    for (int i = 0; i < nRows; ++i) {
        for (int j = i; j < s.length(); j += zig) {
            result  += s[j];
            if (zag < zig && (j + zag) < s.length() && zag > 0)
                result += s[j + zag];
        }
        zag -= 2;
    }
    return result;
}

int main(){
    string s = "PAYPALISHIRING";
    cout << "out: " << convert(s, 3) << endl;
    

    return 0;
}

File: java/convertZigZag.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class convertZigZag {
    public static class Solution {
        public String convert(String s, int nRows) {
            if (nRows == 1 || s.length() <= nRows) return s;
            StringBuffer result = new StringBuffer();
            StringBuffer [] tmp = new StringBuffer[nRows];
            for (int i = 0; i < nRows; i++) 
                tmp[i] = new StringBuffer();
            int start = 0;
            while (start < s.length()) {
                for (int i = 0; i < nRows && start < s.length(); i++) 
                    tmp[i].append(s.charAt(start++));
                for(int i = nRows-2; i >= 1 && start < s.length(); i--)
                    tmp[i].append(s.charAt(start++));
            }
            for (int i = 0; i < nRows; i++) 
                result.append(tmp[i]);
            return result.toString();
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        String s = "PAYPALISHIRING";
        String res = result.convert(s, 3);
        
        System.out.println(res);
    }
}
